File: grade-scores/grade-scores.cpp
// grade-scores.cpp : Defines the entry point for the console application.
//
// This is the "main" component of the TransMax Software Developer Exercise.
//
// Refer to the printUsage(...) function for details.

#include "stdafx.h"
#include <string>

#include <fstream>
#include <iostream>
#include <string>

#include "CResult.hpp"
#include "CResults.hpp"
#include "StringUtils.hpp"

const std::string OUTPUT_FILE_SUFFIX("-graded.txt");

//-----------------------------------------------------------------------------
void printUsage(const std::string& programName)
{
#ifdef _DEBUG
    std::cout << "Usage: " << programName << " <INPUT_FILE>\n"
		"where\n"
		"    <INPUT_FILE> is a .csv file with the format:\n"
		"\n"
		"    <FIRST_NAME>,[SURNAME],<SCORE>\n"
		"\n"
		"       The <FIRSTNAME> field is mandatory.\n"
		"       The [SURNAME] field may be empty (eg. \"Madonna\" goes by one name).\n"
		"       The <SCORE> field is mandatory and must be numeric.\n"
		"       Duplicate <FIRSTNAME>+[SURNAME] combinations are reported and only the\n"
		"       first occurrence is accepted.\n"
		"\n"
		"   Output is written to file <INPUT_FILE>" << OUTPUT_FILE_SUFFIX << " with format:\n"
		"\n"
		"      [SURNAME],<FIRSTNAME>,<SCORE>\n"
		"\n"
		"   The output is sorted by highest <SCORE> and alphabetically on\n"
		"   [SURNAME]+<FIRSTNAME> of students achieving that score.\n";
#endif
}
//-----------------------------------------------------------------------------
void printError(const std::string& programName, const std::string& errorMsg)
{
#ifdef _DEBUG
    std::cout << "ERROR: " << errorMsg << std::endl << std::endl;
	printUsage(programName);
#endif
}

//=============================================================================
// Main
//=============================================================================

int main(int argc, char* argv[])
{
    // Process command line parameters
    //--------------------------------

    const std::string EXECUTABLE_NAME(StringUtils::filenameOf(argv[0]));

    switch (argc)
    {
        case 1:
            printUsage(EXECUTABLE_NAME);
            exit(0);

        case 2:
        {
            // Print usage information if the argument is a "help flag":
            const std::set<std::string> HELP_FLAGS{ "-h","/h","--help","-?","/?" };
            if (HELP_FLAGS.find(argv[1]) != HELP_FLAGS.end())
            {
                printUsage(EXECUTABLE_NAME);
                exit(0);
            }
        }   break;

		default:
			printError(EXECUTABLE_NAME, "Only one input file expected!");
			exit(1);
	}

    // Process the input file, adding all valid entries to the results
    //--------------------=-------------------------------------------

    try
    {
        std::string outputFilePath
                        (StringUtils::newPath(argv[1], OUTPUT_FILE_SUFFIX));
        Results::CResults results(argv[1]);
        results.save(outputFilePath);

#ifdef _DEBUG
        std::cout << "\n+++ Output written to \"" << outputFilePath << "\"\n";
#endif
    }

    catch(Results::InputFileException e)
    {
#ifdef _DEBUG
        std::cerr << "\n*** " << e.what() << std::endl;
#endif
        exit(2);
    }

    catch(Results::OutputFileException e)
    {
#ifdef _DEBUG
        std::cerr << "\n*** " << e.what() << std::endl;
#endif
        exit(3);
    }

	return 0;
}

File: Python3version/grade-scores.py
#!/usr/bin/python3

# TransMax Software Developer Exercise.
#
# Refer to the printUsage(...) function for details.

import os
import sys

def printUsage():
    print('''Usage: %s <INPUT_FILE>
where
  <INPUT_FILE> is a .csv file with the format:
  
      <FIRST_NAME>,[SURNAME],<SCORE>

    The <FIRSTNAME> field is mandatory.
    The [SURNAME] field may be empty (eg. "Madonna" goes by one name).
    The <SCORE> field is mandatory and must be numeric.
    Duplicate <FIRSTNAME>+[SURNAME] combinations are reported and only the
    first occurrence is accepted.
    
Output is written to file <INPUT_FILE>-graded.txt with format:

    [SURNAME],<FIRSTNAME>,<SCORE>
  
The output is sorted by highest <SCORE> and alphabetically on
[SURNAME]+<FIRSTNAME> of students achieving that score.''' % sys.argv[0])
 
def printError(errorMsg):
    print('ERROR: %s\n' % errorMsg)
    printUsage()

if (len(sys.argv) == 1) or (sys.argv[1].lower() in ['-h','/h','--help','?','-?','/?']):
    printUsage()
    
elif len(sys.argv) != 2:
    printError('Only one input file expected!')
    sys.exit(1)

elif not os.path.exists(sys.argv[1]):
    printError('Specified file, "%s", does not exist!' % sys.argv[1])
    sys.exit(2)
    
else:
    # Token indexes of input file fields when each line is tokenised:
    FIRSTNAME = 0
    SURNAME   = 1
    SCORE     = 2
   
    results = {}  # Key=score, Contents=list of student names achieving score
    
    # Populate the results dictionary with the student results, formatting the
    # student names into the required "[SURNAME],<FIRSTNAME>" .csv format
    # to make subsequent sorting easier when outputting the results.
    #
    # Note: For ultimate speed, I could use pure-Python's 
    #       sortedcontainers.SortedDict and sortedcontainers.SortedList but
    #       that's overkill for this exercise.
    #--------------------------------------------------------------------------
    
    try:
        inputFile = open(sys.argv[1])
        lineNo = 1
        studentsAlreadyProcessed = []
        
        for line in inputFile:  # Append all student data to the results dictionary
            
            tokens = line[:-1].replace(' ', '').split(',')  
            # excludes newline character and trims whitespace
            
            if (   (len(tokens) != 3)  # Expected format is <FIRSTNAME>,[SURNAME],<SCORE>
                or (len(tokens[FIRSTNAME]) == 0)  # <FIRSTNAME> is mandatory
                or (len(tokens[SCORE]) == 0)      # <SCORE> is mandatory
                or not tokens[SCORE].isdigit()):  # <SCORE> must be numeric
                print('Ignoring invalid line %i: "%s"' % (lineNo, line[:-1]))
            
            else:
               
                score = int(tokens[SCORE])
                
                # Define thisStudent based on [SURNAME] and <FIRSTNAME>.
                # If the [SURNAME] token is null/empty, use a space character to
                # ensure that it will be the first sorted [SURNAME].
                if len(tokens[SURNAME]) == 0:
                    thisStudent = ' ,%s' % tokens[FIRSTNAME] # " ,<FIRSTNAME>"
                else:
                    thisStudent = '%s,%s' % (tokens[SURNAME], 
                                             tokens[FIRSTNAME]) # "[SURNAME],<FIRSTNAME>"
                
                lowercaseStudentName = thisStudent.lower()
                        
                if lowercaseStudentName in studentsAlreadyProcessed:
                    print('Ignoring duplicate student result at line %i: "%s"' % (lineNo, line[:-1]))
                
                else:  # update results with this student's result

                    if score in results:
                        results[score].append(thisStudent)
                    else:
                        results[score] = [thisStudent]
                        
                    studentsAlreadyProcessed.append(lowercaseStudentName)
                    
            lineNo += 1
        
        inputFile.close()
            
    except IOError:
        printError('Failed to read Results file "%s"!', sys.argv[1])
        sys.exit(3)
        
    # Write the results in .csv format, sorted on score, highest first, then
    # student name, in alphabetical order based on surname, with the score
    # appearing last.
    #-------------------------------------------------------------------------
        
    outputFileName = '%s-graded.txt' % os.path.splitext(sys.argv[1])[0]  
    # Note: "os.path.splitext(sys.argv[1])[0]" excludes the input file's suffix 
    # ie. If the command line's file name is "result.txt", the generated name is
    #     "results-graded.txt" rather than "results.txt-graded.txt"
    
    try:
        outputFile = open(outputFileName, 'w')

        for score in sorted(results.keys(), reverse=True): # process scores, highest first
            for student in sorted(results[score], key=str.lower):  # process sorted student names within this score
                #print('%s,%i' % (student.lstrip(' '), score))
                outputFile.write('%s,%i\n' % (student.lstrip(' '), score))
                # Note: ".lstrip(' ')" is used to eliminate a potential
                #       leading space that was added above to ensure
                #       correct sorting when [SURNAME] is null/empty
        
        outputFile.close()
        
        print('\n+++ Output written to "%s"' % outputFileName)
                    
    except IOError:
        printError('Failed to write file "%s"!', outputFileName)
        sys.exit(4)
